runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (C) 2009-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.grpc.scaladsl
+
+import akka.NotUsed
+import akka.annotation.{ DoNotInherit, InternalApi }
+import akka.grpc.internal.ChannelApiHelpers
+import akka.stream.{ Materializer, OverflowStrategy }
+import akka.stream.scaladsl.{ Sink, Source }
+import io.grpc.stub.{ ClientCalls, StreamObserver }
+import io.grpc.{ CallOptions, Channel, Deadline, MethodDescriptor }
+
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * Request builder for requests giving providing per call metadata capabilities
+ *
+ * Not for user extension
+ */
+@DoNotInherit
+sealed trait RequestBuilder[Req, Res] {
+
+  def addMetadata(key: String, value: String): RequestBuilder[Req, Res]
+  def withDeadline(deadline: FiniteDuration): RequestBuilder[Req, Res]
+  def execute(req: Req): Res
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+final case class RequestBuilderImpl[Req, Res, Ret](
+  options: CallOptions,
+  delegatedExecute: (Req, CallOptions) => Ret) extends RequestBuilder[Req, Ret]() {
+
+  def addMetadata(key: String, value: String): RequestBuilderImpl[Req, Res, Ret] = {
+    // FIXME support values other than string values?
```

Comment:
do 'options' correspond to 'headers' or is there more indirection there?

Reply:
I think there is a bit more indirection, there is also some marshalling infra in the grpc library around this. Not sure if we care about that though, maybe strings are good enough for now, and also avoids tying us down to something client specific. 

It's annotated as an experimental API, but so is most of these CallOptions.